src/audio/module_adapter/module/generic.c

```diff
@@ -347,15 +347,27 @@ int module_free(struct processing_module *mod)
 int module_set_configuration(struct processing_module *mod,
 			     uint32_t config_id,
 			     enum module_cfg_fragment_position pos, size_t data_offset_size,
-			     const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+			     const uint8_t *fragment_in, size_t fragment_size, uint8_t *response,
 			     size_t response_size)
 {
+#if CONFIG_IPC_MAJOR_3
+	struct sof_ipc_ctrl_data *cdata = (struct sof_ipc_ctrl_data *)fragment_in;
+	const uint8_t *fragment = (const uint8_t *)cdata->data[0].data;
+#elif CONFIG_IPC_MAJOR_4
+	const uint8_t *fragment = fragment_in;
+#endif
 	struct module_data *md = &mod->priv;
 	struct comp_dev *dev = mod->dev;
 	size_t offset = 0;
 	uint8_t *dst;
 	int ret;
 
+#if CONFIG_IPC_MAJOR_3
+	if (cdata->cmd != SOF_CTRL_CMD_BINARY) {
+		comp_err(dev, "module_set_configuration(), illegal control command");
+		return -EINVAL;
+	}
+#endif
```

Comment:
It is confusing indeed. In `module_adapter_cmd()` either
`module_adapter_ctrl_get_set_data(dev, cdata, ...);` or
`md->ops->set_configuration(..., cdata,...);` or
`md->ops->get_configuration(..., cdata,...);`
is called. In `module_adapter_ctrl_get_set_data()` then `module_adapter_get_set_params(dev, cdata, set);` and from it
`md->ops->set_configuration(..., (const uint8_t *)cdata->data[0].data,...);` or
`md->ops->get_configuration(..., (const uint8_t *)cdata->data[0].data,...);`
are called, i.e. indeed `.set_configuration()` and `.get_configuration()` are sometimes called with `cdata` and  sometimes with `cdata->data[0].data` as their argument. Isn't the problem already there?

Reply:
Wow indeed, this seems much worse than what comes from a first look. Perhaps a different fix is in order to make everything consistent.

Reply:
@lyakh wait a second

In `module_adapter_ctrl_get_set_data()` then `module_adapter_get_set_params(dev, cdata, set)`; and from it
`md->ops->set_configuration(..., (const uint8_t *)cdata->data[0].data,...);` or
`md->ops->get_configuration(..., (const uint8_t *)cdata->data[0].data,...);`

this doesn't look true to be on my machine.

I see that in `module_adapter_get_set_params(dev, cdata, set)` we call `md->ops->set_configuration(mod, cdata->data[0].type, pos, data_offset_size, (const uint8_t *)cdata, cdata->num_elems,  NULL, 0);` or 

` md->ops->get_configuration(mod, pos, &data_offset_size, (uint8_t *)cdata, cdata->num_elems);`

Reply:
> @lyakh wait a second

@dbaluta ah, you're right, my local tree wasn't updated yet after my vacation, it was still missing 7d0a0c673f94fe59fbe9b7586f841434ac2569b4 which you're actually referencing in your commit description. So, you're right, my mistake!